network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/settings/geoswitching/GeoSwitchingContentProvider.kt

```diff
@@ -16,52 +16,80 @@
+import com.duckduckgo.networkprotection.impl.configuration.WgVpnControllerService
 import com.duckduckgo.networkprotection.impl.settings.geoswitching.GeoSwitchingContentProvider.AvailableCountry
+import com.duckduckgo.networkprotection.store.NetPGeoswitchingRepository
+import com.duckduckgo.networkprotection.store.db.NetPGeoswitchingLocation
 import com.squareup.anvil.annotations.ContributesBinding
 import javax.inject.Inject
+import kotlinx.coroutines.withContext
 
 interface GeoSwitchingContentProvider {
-    fun getContent(): List<AvailableCountry>
+    suspend fun downloadData()
+    suspend fun getDownloadedData(): List<AvailableCountry>
 
     data class AvailableCountry(
         val countryCode: String,
+        val countryName: String,
         val cities: List<String>,
     )
 }
 
-@ContributesBinding(ActivityScope::class)
-class MockGeoSwitchingContentProvider @Inject constructor() : GeoSwitchingContentProvider {
-    override fun getContent(): List<AvailableCountry> {
-        return listOf(
-            AvailableCountry(
-                countryCode = "uk",
-                cities = emptyList(),
-            ),
-            AvailableCountry(
-                countryCode = "fr",
-                cities = emptyList(),
-            ),
-            AvailableCountry(
-                countryCode = "us",
-                cities = listOf("Chicago", "El Segundo", "Newark", "Atlanta"),
-            ),
-            AvailableCountry(
-                countryCode = "ca",
-                cities = emptyList(),
-            ),
-            AvailableCountry(
-                countryCode = "es",
-                cities = listOf("Madrid", "Barcelona"),
-            ),
+@ContributesBinding(AppScope::class)
+class RealGeoSwitchingContentProvider @Inject constructor(
+    private val wgVpnControllerService: WgVpnControllerService,
+    private val dispatcherProvider: DispatcherProvider,
+    private val netPGeoswitchingRepository: NetPGeoswitchingRepository,
+) : GeoSwitchingContentProvider {
+    override suspend fun downloadData() {
+        withContext(dispatcherProvider.io()) {
+            getContent().map {
+                NetPGeoswitchingLocation(
+                    countryCode = it.countryCode,
+                    countryName = it.countryName,
+                    cities = it.cities,
+                )
+            }.also {
+                netPGeoswitchingRepository.replaceLocations(it)
+            }
+        }
+    }
+
+    override suspend fun getDownloadedData(): List<AvailableCountry> = withContext(dispatcherProvider.io()) {
+        netPGeoswitchingRepository.getLocations().map {
             AvailableCountry(
-                countryCode = "de",
-                cities = emptyList(),
-            ),
+                countryCode = it.countryCode,
+                countryName = it.countryName,
+                cities = it.cities,
+            )
+        }
+    }
+
+    private suspend fun getContent(): List<AvailableCountry> {
```

Comment:
Similar naming comment, getContent is very generic and `AvailableCountry` is also more like `ServerLocation` ?

Reply:
Updated `AvailableCountry` to `ServerLocation`